compiler/src/iree/compiler/Codegen/LLVMGPU/KernelConfig.cpp

```diff
@@ -260,6 +265,218 @@ static CodeGenPipeline getTensorCorePipeline(Type elementType) {
+  return isa<linalg::ReduceOp>(op) ||
+         (isa<linalg::GenericOp>(op) &&
+          llvm::any_of(op.getIteratorTypesArray(),
+                       linalg::isReductionIterator));
+}
+
+// Get the bitwidth of the operation.
+static llvm::FailureOr<int64_t> getBitWidth(linalg::LinalgOp op) {
+
+  Value init = op.getDpsInitOperand(0)->get();
+  Value src = op.getDpsInputOperand(0)->get();
+  Type initElemType = getElementTypeOrSelf(init);
+  Type srcElemType = getElementTypeOrSelf(src);
+
+  if (auto initOp = init.getDefiningOp<linalg::GenericOp>()) {
+    if (IREE::LinalgExt::isBitExtendOp(initOp))
+      initElemType = getElementTypeOrSelf(initOp.getDpsInputs()[0]);
+  }
+
+  if (auto srcOp = src.getDefiningOp<linalg::GenericOp>()) {
+    if (IREE::LinalgExt::isBitExtendOp(srcOp))
+      srcElemType = getElementTypeOrSelf(srcOp.getDpsInputs()[0]);
+  }
+
+  if (!initElemType.isIntOrFloat() || !srcElemType.isIntOrFloat())
+    return failure();
+
+  int64_t bitWidth = std::min(initElemType.getIntOrFloatBitWidth(),
+                              srcElemType.getIntOrFloatBitWidth());
+
+  // Reduction distribution only supports 8/16/32 bit types now.
+  if (!llvm::is_contained({8, 16, 32}, bitWidth))
+    return failure();
+
+  return bitWidth;
+}
+
+/// The kernel config is for a single reduction op within a
+/// dispatch. The reduction op shouldn't have any consumer
+/// because that may introduce new shared dimensions and the
+/// distribution analysis will fail. The lowering config is
+/// only attached to the reduction op for now.
+/// TODO(pashu123): Analyze the shared parallel and reduction
+/// dimensions within a dispatch. Based on the analysis try to
+/// put the lowering_config on each of the operation within a
+/// dispatch. That way we can support multiple reduction ops
+/// within a dispatch.
+static LogicalResult
+setReductionVectorDistributionConfig(IREE::GPU::TargetAttr target,
+                                     mlir::FunctionOpInterface entryPoint,
+                                     linalg::LinalgOp op) {
+
+  int64_t countReduction = 0;
+
+  // TODO(pashu123): Remove this check and allow multiple reductions.
+  WalkResult walkResult = entryPoint.walk([&](Operation *op) {
+    if (auto linalgOp = dyn_cast<linalg::LinalgOp>(op)) {
+      if (hasReductionIterator(linalgOp)) {
+        countReduction++;
+      }
+    }
+    if (countReduction > 1) {
+      return WalkResult::interrupt();
+    }
+    return WalkResult::advance();
+  });
+
+  if (walkResult.wasInterrupted()) {
+    return failure();
+  }
+
+  // TODO(pashu123): The consumer operation may introduce new
+  //  dimension which will fail the distribution analysis.
+  for (auto potentialLinalgUser : op->getUsers()) {
+    if (auto linalgOp = dyn_cast<linalg::LinalgOp>(potentialLinalgUser)) {
+      return failure();
+    }
+  }
+
+  SmallVector<unsigned> parallelDims;
+  SmallVector<unsigned> reductionDims;
```

Comment:
nit: don't use unsigned (if possible)

Reply:
IIRC, this is the type used in various helpers for dims, so I wouldn't deviate from that

Reply:
ah ok